File: player.py
from board import Board
from ship import SmallL, ThreePieceShip


class Player:

    def __init__(self, name) -> None:
        self.board = Board()
        self.hit_tracker = [[0 for i in range(10)] for j in range(10)]
        self.name = name
        self.ships = [SmallL(), SmallL(), SmallL(), SmallL()]
    
    def choose_coordinates(self, coordinates):
        x, y = coordinates
        self.hit_tracker[y][x]
        if self.hit_tracker[y][x] == 0:
            return coordinates
    
    def check_lost(self):
        return self.board.check_game_over()
    
    def cli_choose_coordinates(self):
        x = int(input("x: "))
        y = int(input("y: "))
        return self.choose_coordinates((x,y))

    def place_missile(self, coordinates):
        result = self.board.place_missile(coordinates)
        return result

    def update_hit_tracker(self, coordinates):
        for key, value in coordinates.items():
            x, y = key
            self.hit_tracker[y][x] = value

    def print_hit_tracker(self):
        for i in range(len(self.hit_tracker)):
            print(self.hit_tracker[i])

    def place_ships(self, ship_nr, rotation, coordinate):
        if ship_nr < len(self.ships):
            ship = self.ships[ship_nr]
            if ship.placed == False:
                if rotation > 0:
                    ship.rotate_right(rotation)
                self.board.place_ship(coordinate, ship)
            else:
                print("NOT A VALID SHIP PLACEMENT")
        else:
            print("NOT A VALID SHIP")

    def all_ships_placed(self):
        for ship in self.ships:
            if ship.placed == False:
                print("place all ships")
                return False
        return True

    def print_ships_not_placed(self):
        print("ships:", end="\t")
        for i, ship in enumerate(self.ships):
            if ship.placed == False:
                print(f"{i}", end="\t")
        print()

File: server/game_logic/player.js
import { Board } from "./board.js";
import { SmallLShip } from "./ship.js";

export class Player{

    constructor(name){
        this.name = name;
        this.hitTracker = this.initHitTracker();
        this.board = new Board()
        this.ships = [new SmallLShip(), new SmallLShip(), new SmallLShip()]
    }

    placeMissile(coordinate){
        result = this.board.placeMissile(coordinate)
        return result;
    }

    updateHitTracker(coordinates){ // map of what coordinate(string) and value
        for (let coordinate of coordinates){
            var stringCoordinate = coordinate[0];
            var value = coordinate[1];
            var x = parseInt(stringCoordinate[0]);
            var y = parseInt(stringCoordinate[1]);
            this.hitTracker[y][x] = value;
        }
    }

    printHitTracker(){
        console.log(this.hitTracker);
    }

    placeShip(coordinate, shipNumber){
        console.log(coordinate)
        if (shipNumber < this.ships.length){
            var ship = this.ships[shipNumber];
            if (this.board.placeShip(coordinate, ship)){
                return {"x": coordinate.x, "y":coordinate.y, "shipNumber": shipNumber};
            } else {
                console.log("not a valid ship placement");
                return {}
            }
        } else {
            console.log("not a valid ship");
            return {}
        }
    }

    allShipsPlaced(){
        for (var i = 0; i < this.ships.length; i++){
            if (this.ships[i].placed == false){
                print("place all ships")
            }
            return false;
        }
        return true;
    }

    printShipsNotPlaced(){
        for (var i = 0; i < this.ships.length; i++){
            if (this.ships[i].placed === false){
                console.log(i)
            }
        }
        console.log()
    }

    rotateShip(shipNumber){
        var ship = this.ships[shipNumber];
        var new_offsets = ship.rotateRight(1);
        return new_offsets;
    }

    getBoard(){
        return this.board.board;
    }

    getHitTracker(){
        return this.hitTracker;
    }

    getShipsOffsets(){
        var shipsOffsets = [];
        for (var i = 0; i < this.ships.length; i++){
            shipsOffsets.push(this.ships[i].offsets);
        }
        return shipsOffsets;
    }

    initHitTracker(){
        var hitTracker = []
        for (var i = 0; i < 10; i++){
            var row = []
            for (var j = 0; j < 10; j++){
                row.push(0)
            }
            hitTracker.push(row)
        }
        return hitTracker
    }
}
